File: backend/app/models/enrollment.py
from datetime import date, datetime
from typing import TYPE_CHECKING

from sqlalchemy import Date, DateTime, Enum, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.enums import EnrollmentStatus
from app.database.base import Base

if TYPE_CHECKING:
    from app.models.plan import Plan
    from app.models.student import Student


class Enrollment(Base):
    __tablename__ = "enrollments"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), index=True, nullable=False)
    plan_id: Mapped[int] = mapped_column(ForeignKey("plans.id"), index=True, nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[EnrollmentStatus] = mapped_column(
        Enum(EnrollmentStatus, native_enum=False, length=20),
        nullable=False,
        default=EnrollmentStatus.ACTIVE,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    student: Mapped["Student"] = relationship()
    plan: Mapped["Plan"] = relationship()

File: frontend/src/types/enrollment.ts
import type { PageQueryParams } from "./common";
import type { PaymentStatus } from "./payment";

export type EnrollmentStatus = "ACTIVE" | "EXPIRED" | "CANCELLED";

export interface Enrollment {
  id: number;
  student_id: number;
  plan_id: number;
  start_date: string;
  end_date: string;
  status: EnrollmentStatus;
  payment_status: PaymentStatus | null;
  created_at: string;
  updated_at: string;
}

export interface EnrollmentCreatePayload {
  student_id: number;
  plan_id: number;
  start_date: string;
  first_payment_due_date?: string | null;
}

export interface EnrollmentUpdatePayload {
  start_date?: string;
  end_date?: string;
  status?: EnrollmentStatus;
}

export interface EnrollmentQueryParams extends PageQueryParams {
  student_id?: number;
  plan_id?: number;
  status?: EnrollmentStatus;
  student_search?: string;
}
